### PR TITLE
COMPASS-542: Build installers for Ubuntu and Redhat

### PR DESCRIPTION
**Pull request by imlucas**
*note* `npm version major` after merging.

## Todo

- [x] Build .rpm's for RHEL 
- [x] Build .deb's for Ubuntu
- [x] Build tar.gz
- [x] Get it all working on travis
- [x] Proper icon support 
- [x] Add `slug` to Config class (`_.dasherize('${name}-${channel suffix}')` to pass to `electron-installer-debian` and `electron-installer-redhat`.
- [x] `electron-installer-debian` and `electron-installer-redhat` produce wacky filenames. Figure it out and add tests to make sure they're producing the right files.
- [x] Travis failing rpm build because of `error: line 2: Illegal char '-' in: Version: 1.2.0-dev`

#### App icons for Linux

- Convert `.icns` to a `.png` set (e.g. https://iconverticons.com/online/) 
- Update package.json hadron build config for icon path
- Pass `icon` option for all [`BrowserWindow`](http://electron.atom.io/docs/api/browser-window/#new-browserwindowoptions) and [`dialog`](http://electron.atom.io/docs/api/dialog/#dialogshowmessageboxbrowserwindow-options-callback) calls

#### User experience check for rpms (Sam)

> As a redhat user, I want to feel like the `.rpm`'s are native to my experience so that I can trust the app I'm installing.

#### User experience check for debs (Peter)

> As an Ubuntu user, I want to feel like the `.deb`'s are native to my experience so that I can trust the app I'm installing.

> `synopsis` generated from description (as our description is also <80 characters, though generally in the same style as a git commit, i.e. read until the `\r\n\r\n` or 79 characters, whichever comes first) would be nice here: https://www.debian.org/doc/debian-policy/ch-binary.html#s-synopsis. Have not looked at any other options though, hopefully lintian helps there.

#### Codesign all linux artifacts

The notary-service already has support for `.zip`, `.tar.gz`, `.deb` and `.rpm`. Write a js api client so that the `release` command will produce codesigned artifacts as we do for all other platforms.

https://github.com/10gen/notary-service

https://github.com/10gen/notary-service/tree/master/client

Current state of signing:

- macOS: The notary-service doesn't support signing `.app's`. We have a special evergreen distro that has personal key on it. 
- windows: Uses the notary-service. [More details](https://jira.mongodb.org/browse/BUILD-920)

#### Get it all working on evergreen

Already [a ticket for ubuntu test failure](https://jira.mongodb.org/browse/COMPASS-479), but some additional risk in missing packages, distro availability, etc.

### Backlog

Other distribution formats we may want to/have requests for:

- [AppImage](https://github.com/electron-userland/electron-builder/blob/master/src/targets/appImage.ts) assets
- [Snap](https://github.com/electron-userland/electron-builder/blob/master/src/targets/snap.ts) assets
- Arch Linux pacman https://github.com/Filirom1/npm2arch https://github.com/jordansissel/fpm/blob/master/lib/fpm/package/pacman.rb

## Notes

- [Deps for building rpm's](https://github.com/unindented/electron-installer-redhat#requirements)
- [Deps for building deb's (only required when building on macOS)](https://github.com/unindented/electron-installer-debian#requirements)
- [Lintian](https://lintian.debian.org/) looks interesting
  > dissects Debian packages and tries to find bugs and policy violations. It contains automated checks for many aspects of Debian policy as well as some checks for common errors.
- [How Atom creates rpms](https://github.com/atom/atom/blob/master/script/lib/create-rpm-package.js)
- [How Atom creates debs](https://github.com/atom/atom/blob/master/script/lib/create-debian-package.js)
- [Atom's desktop.ini and other spec files](https://github.com/atom/atom/tree/master/resources/linux)
- How to convert semver version string to deb/rpm compliant?
  - From: https://github.com/mojombo/semver/issues/145#issuecomment-195549847
    > But in the .deb and .rpm I have to make concessions to those systems. In particular, it's easy enough for a .deb to have version 1.0.0, but for 2.0.0-rc.1, I'll use 2.0.0~rc.1. This mapping is bijective.
    > Similarly, in rpm, I can't have any separator other than a period. Again, perfectly fine for stable releases, but for prerelease versions, they say to use the release field. So 2.0.0-rc.1 will instead become VERSION=2.0.0 and RELEASE=0.1.rc.1 (in particular, I plan on doing RELEASE=0.<n>.rc.<n> for release candidate n).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -246,25 +249,37 @@ Which assets are generated depends on the target platform.
...
+- `.rpm`: Installer for Redhat
+- `.deb`: Installer for Ubuntu and Debian
+- `.tar.gz`: Convenience for easy automation on unix
+- `-linux-${arch}.zip`: Convenience for easy automation on win32-x64
+
 ## Auto Update
 
 - `hadron-auto-update-manager`: https://github.com/hadron-auto-update-manager
 
+### macOS
```
Comment (pzrq) [inline README.md]: This branch has grown large, would it be reasonable to extract the macOS / Windows / electron-prebuilt changes and get those merged independently?

File: commands/config.js [hunk 2]
```diff
@@ -12,23 +12,106 @@ exports.command = 'config';
...
+    description: {
+      describe: 'What is the description of the application we should display to humans?',
+      default: process.env.npm_package_description || pkg.description
+    },
+    sign: {
+      describe: 'Should this build be signed?',
+      type: 'boolean',
+      default: true
+    },
+    signtool_params: {
```
Comment (pzrq) [inline commands/config.js]: Is this only for electron-winstaller, or does it apply to electron-installer-codesign and/or electron-osx-sign?

File: commands/config.js [hunk 3]
```diff
@@ -53,6 +136,7 @@ const toTable = (CONFIG) => {
 };
 
 exports.handler = (argv) => {
+  console.log('config argv', argv);
```
Comment (pzrq) [inline commands/config.js]: Does eslint catching these in hadron-build?

File: commands/config.js [hunk 4]
```diff
@@ -12,23 +12,106 @@ exports.command = 'config';
...
+      default: pkg.github_owner
+    },
+    github_repo: {
+      default: pkg.github_repo
+    },
+    author: {
+      default: pkg.author || pkg.authors
+    }
+  };
+  _.assign(opts, {
```
Comment (pzrq) [inline commands/config.js]: Should we continue using `_.assign` over `Object.assign`? We tend to get warnings when installing packages, not sure if they are related:

screen shot 2016-12-13 at 10 42 05 pm

File: lib/config.js [hunk 5]
```diff
@@ -408,21 +356,120 @@ exports.get = (cli, callback) => {
...
+        function(cb) {
+          which('fakeroot', function(err) {
+            if (err) {
+              cli.warn('Your environment is not configured correctly to build ' +
+              'debian packages. Please see https://git.io/v1iRV');
+              return cb();
+            }
+            /**
+             * TODO (imlucas) Use pretty debian metadata and options.
+             * @see https://github.com/unindented/electron-installer-debian#options
```
Comment (pzrq) [inline lib/config.js]: `synopsis` generated from description (as our description is also <80 characters, though generally in the same style as a git commit, i.e. read until the `\r\n\r\n` or 79 characters, whichever comes first) would be nice here: https://www.debian.org/doc/debian-policy/ch-binary.html#s-synopsis

Have not looked at any other options though, hopefully lintian helps there.

File: .travis.yml [hunk 6]
```diff
@@ -1,8 +1,19 @@
 sudo: false
+dist: trusty
 language: node_js
 node_js:
   - 6
 script: npm run-script ci
 cache:
+  # TODO (imlucas) We can't use yarn just yet on travis until an upstream
+  # fix is landed for https://github.com/yarnpkg/yarn/pull/2164
```
Comment (imlucas) [inline .travis.yml]: Yes but `0.19.0` is far from being published.

File: lib/config.js (hunk 5 above)
Comment (imlucas) [inline lib/config.js]: 👍 will include in follow-up tickets.

File: commands/config.js (hunk 2 above)
Comment (imlucas) [inline commands/config.js]: `electron-winstaller` only. This will be cleaned up as the refactoring to `Target`s instead of `CONFIG` makes progress.

**Pull request merged**